src/dag_algo/mod.rs

```diff
@@ -637,3 +642,86 @@ pub fn collect_bicolor_runs(
 
     Ok(block_list)
 }
+
+/// Returns the transitive reduction of a directed acyclic graph
+///
+/// The transitive reduction of G = (V,E) is a graph G- = (V,E-)
+/// such that for all v,w in V there is an edge (v,w) in E- if and only if (v,w) is in E
+/// and there is no path from v to w in G with length greater than 1.
+///
+/// :param PyDiGraph graph: A directed acyclic graph
+///
+/// :returns: a directed acyclic graph representing the transitive reduction, and a map containing the index of a node in the original graph mapped to its equivalent in the resulting graph.
+/// :rtype: Tuple[:class:`~rustworkx.PyGraph`, dict]
+///
+/// :raises PyValueError: if graph is not a DAG
+
+#[pyfunction]
+#[pyo3(text_signature = "(graph, /)")]
+pub fn transitive_reduction(
+    graph: &digraph::PyDiGraph,
+    py: Python,
+) -> PyResult<(digraph::PyDiGraph, DictMap<usize, usize>)> {
+    let g = &graph.graph;
+    let mut index_map = DictMap::new();
+    if !is_directed_acyclic_graph(graph) {
+        return Err(PyValueError::new_err(
+            "Directed Acyclic Graph required for transitive_reduction",
+        ));
+    }
+    let mut tr = StablePyGraph::<Directed>::new();
+    let mut descendants = DictMap::new();
+    let mut check_count = DictMap::new();
```

Comment:
I'd probably use `HashMap` for this, because `DictMap` is only advantageous if you need to return it to python as it retains insertion order for iteration (or if you need to iterate over it). But the use here is only for lookups/edits by key so a `HashMap` will be faster.

```suggestion
    let mut check_count = HashMap::with_capacity(g.node_count());
```